src/js/color.ts

```diff
@@ -0,0 +1,3845 @@
+ * @returns {string} - hex color
+ */
+export const convertXyzToHex = (xyz: Array<number>): string => {
+  const [x, y, z, alpha] = validateColorComponents(xyz, {
+    minLength: QUAT,
+    validateRange: false,
+  })
+  const [r, g, b] = transformMatrix(MATRIX_XYZ_TO_L_RGB, [x, y, z], true)
+  const hex = convertLinearRgbToHex(
+    [
+      Math.min(Math.max(r!, 0), 1),
+      Math.min(Math.max(g!, 0), 1),
+      Math.min(Math.max(b!, 0), 1),
+      alpha,
+    ],
+    true,
+  )
+  return hex
+}
+
+/**
+ * convert xyz D50 to hex color
+ * @param {Array.<number>} xyz - [x, y, z, alpha]
+ * @returns {string} - hex color
+ */
+export const convertXyzD50ToHex = (xyz: Array<number>): string => {
+  const [x, y, z, alpha] = validateColorComponents(xyz, {
+    minLength: QUAT,
+    validateRange: false,
+  })
+  const xyzD65 = transformMatrix(MATRIX_D50_TO_D65, [x, y, z], true)
+  const [r, g, b] = transformMatrix(MATRIX_XYZ_TO_L_RGB, xyzD65, true)
+  const hex = convertLinearRgbToHex([
+    Math.min(Math.max(r!, 0), 1),
+    Math.min(Math.max(g!, 0), 1),
+    Math.min(Math.max(b!, 0), 1),
+    alpha,
+  ])
+  return hex
+}
+
+/**
+ * convert xyz to rgb
+ * @param {Array.<number>} xyz - [x, y, z, alpha]
+ * @param {boolean} skip - skip validate
+ * @returns {Array.<number>} - [r, g, b, alpha] r|g|b: 0..255 alpha: 0..1
+ */
+export const convertXyzToRgb = (
+  xyz: Array<number>,
+  skip: boolean = false,
+): Array<number> => {
+  let x, y, z, alpha
+  if (skip) {
+    ;[x, y, z, alpha] = xyz
+  } else {
+    ;[x, y, z, alpha] = validateColorComponents(xyz, {
+      validateRange: false,
+    })
+  }
+  let [r, g, b] = transformMatrix(MATRIX_XYZ_TO_L_RGB, [x, y, z], true)
+  ;[r, g, b] = convertLinearRgbToRgb(
+    [
+      Math.min(Math.max(r!, 0), 1),
+      Math.min(Math.max(g!, 0), 1),
+      Math.min(Math.max(b!, 0), 1),
+    ],
+    true,
+  )
+  return [r, g, b, alpha]
+}
+
+/**
+ * convert xyz to xyz-d50
+ * @param {Array.<number>} xyz - [x, y, z, alpha]
+ * @returns {Array.<number>} - [x, y, z, alpha]
+ */
+export const convertXyzToXyzD50 = (xyz: Array<number>): Array<number> => {
+  const [xx, yy, zz, alpha] = validateColorComponents(xyz, {
+    validateRange: false,
+  })
+  const [x, y, z] = transformMatrix(MATRIX_D65_TO_D50, [xx, yy, zz], true)
+  return [x, y, z, alpha]
+}
+
+/**
+ * convert xyz to hsl
+ * @param {Array.<number>} xyz - [x, y, z, alpha]
+ * @param {boolean} skip - skip validate
+ * @returns {Array.<number>} - [h, s, l, alpha]
+ */
+export const convertXyzToHsl = (
+  xyz: Array<number>,
+  skip: boolean = false,
+): Array<number> => {
+  const [rr, gg, bb, alpha] = convertXyzToRgb(xyz, skip)
+  const r = rr! / MAX_RGB
+  const g = gg! / MAX_RGB
+  const b = bb! / MAX_RGB
+  const max = Math.max(r, g, b)
+  const min = Math.min(r, g, b)
+  const d = max - min
+  const l = (max + min) * HALF * MAX_PCT
+  let h, s
+  if (Math.round(l) === 0 || Math.round(l) === MAX_PCT) {
+    h = NONE as never
```

Comment:
NONE is for a CSS keyword 'none', a string.
Maybe it's because I'm not very familiar with TypeScript, what's the intent of this `never`?

Reply:
TLDR is that the return type is `Array<number>` and we need to keep this type a number so the `as never` just ignores the string type for now.

This type cast _should_ be removed at some point in the future when the return types are more accurate to the code usage.

Reply:
Got it.